Cython/Compiler/PyrexTypes.py

```diff
@@ -335,6 +335,10 @@ def is_simple_buffer_dtype(self):
         return (self.is_int or self.is_float or self.is_complex or self.is_pyobject or
                 self.is_extension_type or self.is_ptr)
 
+    def is_optional_type(self):
+        """Returns True if type can be used with typing.Optional[]."""
+        return (self.is_pyobject or self.is_buffer or self.is_memoryviewslice)
+
```

Comment:
Two comments here. First, this is a method on type classes, so `_type` in the name is redundant. That's also why I proposed `can_be_optional` as name, since `is_optional` suggests that `Optional` was actually used on the type.

Secondly, the base class shouldn't know which types can be optional. Only the concrete types can take the decision. Thus, this method should return `False` here and subclasses that allow `Optional[]` should override it and return `True` (or take a more elaborate decision, but they'll probably just return a constant as it stands).

I just noticed that the second argument equally applies to the `is_simple_buffer_dtype()` method above. It also shouldn't know all types that it applies to. I'll fix that.

Reply:
> why I proposed can_be_optional as name

This is mistake. I will change it to `can_be_optional`

Reply:
Done. But not sure if the position of method is correct (if there is an ordering of methods in classes or not).